From cc3b5f3a42141dbec03f39f7d1080224f9a1a2c7 Mon Sep 17 00:00:00 2001
From: Ingo Molnar <mingo@elte.hu>
Date: Fri, 3 Jul 2009 08:30:27 -0500
Subject: [PATCH 131/309] stop_machine: convert stop_machine_run() to
 PREEMPT_RT

Instead of playing with non-preemption, introduce explicit
startup serialization. This is more robust and cleaner as
well.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
[bigeasy: XXX: stopper_lock -> stop_cpus_lock]
---
 kernel/stop_machine.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/kernel/stop_machine.c b/kernel/stop_machine.c
index 01fbae5b97b7..6c44ff1413a5 100644
--- a/kernel/stop_machine.c
+++ b/kernel/stop_machine.c
@@ -466,6 +466,16 @@ repeat:
 		struct cpu_stop_done *done = work->done;
 		char ksym_buf[KSYM_NAME_LEN] __maybe_unused;
 
+		/*
+		 * Wait until the stopper finished scheduling on all
+		 * cpus
+		 */
+		lg_global_lock(&stop_cpus_lock);
+		/*
+		 * Let other cpu threads continue as well
+		 */
+		lg_global_unlock(&stop_cpus_lock);
+
 		/* cpu stop callbacks are not allowed to sleep */
 		preempt_disable();
 
-- 
1.8.1.2

